src/components/AppSettingsModal/AppSettingsModal.js

```diff
@@ -45,29 +46,13 @@ class AppSettingsModal extends PureComponent<Props, State> {
     saveAppSettings(newSettings);
   };
 
-  // todo: refactor state update methods into single method & use dotty https://www.npmjs.com/package/dotty
-  //       method params keyString, value
-  //       --> keyString = general.defaultProjectPath
-  selectDefaultProjectPath = (selectedPath): AppSettings => {
-    this.setState(
-      produce(draftState => {
-        draftState.newSettings.general.defaultProjectPath = selectedPath;
-      })
-    );
-  };
-
-  selectDefaultProjectType = (projectType: ProjectType): AppSettings => {
-    this.setState(
-      produce(draftState => {
-        draftState.newSettings.general.defaultProjectType = projectType;
-      })
-    );
-  };
-
-  toggleUsageTracking = (enableUsageTracking: boolean): AppSettings => {
-    this.setState(
-      produce(draftState => {
-        draftState.newSettings.privacy.enableUsageTracking = enableUsageTracking;
+  // Update setting by key
+  // Notice: Key will be nested e.g. 'general.defaultProjectPath' that's why we're using utility function setNested
+  updateSetting = (key: string, value: any) => {
+    this.setState(state =>
+      // produce could be used with currying here - so no state passed to produce. But Flow couldn't detect the right return type.
+      produce(state, draftState => {
+        setNested(draftState.newSettings, key, value);

```

Comment:
hum I understand why you want this change but this is not typed at all... You could pass any key and it will just set it. I think I prefer the previous version, more secure.
I'd say you don't need to nest the settings, that's just a UI thing. If you have a flat object you can then do something like
```
updateSetting = (key: $Keys<AppSettings>, value: any) => {
}
```
It'd be nice to type the value as well instead of `any` but that's a lot harder. That's why I'm really not sure about this entire change. Having 3 separate functions suits me fine tbh

Reply:
Yes, you're right there is no type checking and that's bad. But I'm not sure if we need it. I think we could generate the settings modal dynamically based on the `appSettings` state later.
That's also the reason why the settings are nested in state. 

Anyway I could revert to the three methods but I like that it's DRY here and it should be pretty easy to map over the settings to create the modal.